releng_sop/koji_sign.py

```diff
@@ -0,0 +1,715 @@
+from __future__ import print_function
+
+import base64
+import logging
+import multiprocessing.dummy
+import os
+import shutil
+import subprocess
+import tempfile
+
+import koji
+
+from .common import get_logger
+
+
+__all__ = (
+    "KojiSignRPMs",
+    "get_rpmsign_class",
+)
+
+
+class KojiSignRPMs(object):
+    """
+    Sign RPMs and import them to a koji instance.
+
+    :param koji_profile: Koji profile name
+    :type  koji_profile: str
+    :param rpmsign_class: Reference to a class that implements signing
+    :type  rpmsign_class: object
+    :param logger: Custom logger
+    :type  logger: logging.Logger
+    :param log_level: Log level for default logger (when logger is not set)
+    :type  log_level: int
+    """
+
+    def __init__(self, koji_profile, rpmsign_class, logger=None, log_level=logging.INFO):  # noqa: D102
+        self.koji_profile = koji_profile
+        self.koji_module = koji.get_profile_module(self.koji_profile)
+        self.koji_session = koji.ClientSession(self.koji_module.config.server)
+        self.rpmsign_class = rpmsign_class
+        self._get_rpm_sighdr_sigkey_cache = {}
+        self.logger = logger or get_logger(self, log_level)
+
+        if self.koji_module.config.authtype == "kerberos":
+            self.koji_session.krb_login()
+
+    def log(self, level, msg, commit=False):
+        """
+        Logging wrapper that prepeds [TEST] if commit is False.
+
+        :param level: Logging level (info, debug, ...)
+        :type  level: str
+        :param msg: Message
+        :type  msg: str
+        :param commit: Disable dry-run, apply changes for real.
+        :type  commit: bool=False
+        """
+        func = getattr(self.logger, level)
+        if not commit:
+            msg = "[TEST] %s" % msg
+        func(msg)
+
+    def get_latest_tagged_rpms(self, tag_name, inherit=False, packages=None):
+        """
+        Return rpm_info list for latest tagged builds in a tag.
+
+        :param tag_name: A koji tag name
+        :type  tag_name: str
+        :param inherit: Follow tag inheritance
+        :type  inherit: bool=False
+        :param packages: Filter RPMs by provided packages
+        :type  packages: list=None
+        :return: List of koji rpm_info dictionaries
+        :rtype:  list
+        """
+        result = []
+
+        if not packages:
+            packages = [None]
+
+        for package in set(packages):
+            rpm_info_list, build_info_list = self.koji_session.listTaggedRPMS(tag_name, latest=True, inherit=inherit, package=package, rpmsigs=False)
+            builds_by_id = {}
+            for build_info in build_info_list:
+                build_info["name"] = build_info["package_name"]
+                builds_by_id[build_info["id"]] = build_info
+            for rpm_info in rpm_info_list:
+                rpm_info["build"] = builds_by_id[rpm_info["build_id"]]
+            result.extend(rpm_info_list)
+        return result
+
+    def get_build_rpms(self, build_list):
```

Comment:
That's ok, it's prepared for a different script or a user using the module as a library.